### DISCUSSION TITLE
Unable to compile hls model

### DISCUSSION THREAD

**Discussion by Sam-Nielsen-Dot** [fastmachinelearning/hls4ml#672] [Q&A]
When I try to run the first tutorial notebook from hls4ml-tutorial

The code stops at the hls_model.compile() line and shows the error

```
Writing HLS project
Done
build_lib.sh: line 14: g++: command not found
build_lib.sh: line 15: g++: command not found
build_lib.sh: line 16: g++: command not found
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
Input In [13], in <cell line: 1>()
----> 1 hls_model.compile()
      2 X_test = np.ascontiguousarray(X_test)
      3 y_hls = hls_model.predict(X_test)

File ~/anaconda3/envs/hls4ml-tutorial/lib/python3.8/site-packages/hls4ml/model/hls_model.py:546, in HLSModel.compile(self)
    544         dlclose_func.restype = ctypes.c_int
    545         dlclose_func(self._top_function_lib._handle)
--> 546     self._top_function_lib = ctypes.cdll.LoadLibrary(lib_name)
    547 finally:
    548     os.chdir(curr_dir)

File ~/anaconda3/envs/hls4ml-tutorial/lib/python3.8/ctypes/__init__.py:451, in LibraryLoader.LoadLibrary(self, name)
    450 def LoadLibrary(self, name):
--> 451     return self._dlltype(name)

File ~/anaconda3/envs/hls4ml-tutorial/lib/python3.8/ctypes/__init__.py:373, in CDLL.__init__(self, name, mode, handle, use_errno, use_last_error, winmode)
    370 self._FuncPtr = _FuncPtr
    372 if handle is None:
--> 373     self._handle = _dlopen(self._name, mode)
    374 else:
    375     self._handle = handle

OSError: firmware/myproject-478C5A15.so: cannot open shared object file: No such file or directory
```

This same error occurs when I try to run any of the other tutorial notebooks or compile a HLS4ML model in any other piece of code

Furthermore,

when I try to just build the model without compiling it I get the error message

Project myproject_prj does not exist. Rerun "hls4ml build -p my-hls-test".

Do you know how to rectify this?

Thankyou

**Comment by vloncar** [comment]
Some tips on solving that can be found [here](https://gprivate.com/60jqj).

**Reply by AnthonyAportela** [reply]
What ends up happening when you give a reply like this is that it shows up as a google result.

**Comment by adityaroy10** [comment]
Hello, I have been facing the same error. Did you figure out the solution?

**Reply by jmitrevs** [reply]
You need to have g++ installed.

**Reply by adityaroy10** [reply]
I have g++ installed. I am trying to compile my own model.

I should have clarified this earlier. I am encountering the same OSerror.
Additionally, I can run the tutorial successfully. I am encountering this error only when trying to compile my model.

Here is a screenshot of the error

![image](https://github.com/fastmachinelearning/hls4ml/assets/69969576/9bb3a43d-64cb-4b68-935c-48b3a01fb771)
![image](https://github.com/fastmachinelearning/hls4ml/assets/69969576/539f5475-6763-45f9-b6d5-6b12a19ebc26)

**Reply by Govarshini99** [reply]
Same error i'm facing 
have you solved this issue by any chance , i facing this only when i compile my own model.

**Reply by Govarshini99** [reply]
![Screenshot from 2024-09-27 20-05-08](https://github.com/user-attachments/assets/0ddee5d6-ccaa-4408-9402-687ad0eb9788)

**Reply by jmitrevs** [reply]
In this case hls4ml is reporting that it can't open the library, most likely because it didn't exist. Can you look further back when the library was being compiled to see if there are errors?

**Reply by Govarshini99** [reply]
I didn't have any errors before this

**Reply by Govarshini99** [reply]
I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-09-27 22:13:53.108362: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-09-27 22:13:53.542386: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2024-09-27 22:14:06.907216: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-09-27 22:14:06.930246: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1960] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2024-09-27 22:14:15.650756: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x71ec680086d0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2024-09-27 22:14:15.650783: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2024-09-27 22:14:15.680605: I ./tensorflow/compiler/jit/device_compiler.h:186] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.

got these set of warnings during during evaluate step before using hls converter

**Reply by Govarshini99** [reply]
but this is related to gpu usage

**Comment by eason0307** [comment]
I also have the same problem
It is the first example code
My OS is RHEL9.4
![Screenshot from 2024-09-28 08-12-16](https://github.com/user-attachments/assets/0cfb4b76-ca27-4097-982f-acefb794b102)
Does it have the solution?

**Reply by jmitrevs** [reply]
You need to look for compilation errors. What you are showing is the effect, not the cause. Note, if this is on Jupyter the compilation errors might appear on the terminal where you started Jupyter.

**Reply by eason0307** [reply]
![Screenshot from 2024-09-29 04-06-16](https://github.com/user-attachments/assets/c9248d50-426e-4b55-90bc-e135a391353c)
My terminal is showing this error message when execute compile 
But when I using Ubuntu doesn't have any error

**Reply by Govarshini99** [reply]
g++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
/usr/bin/ld: cannot find myproject.o: No such file or directory
collect2: error: ld returned 1 exit status

My terminal is showing this error

**Reply by vloncar** [reply]
@eason0307 and @Govarshini99 Google this error for a list of potential causes and solutions. There are a few discussions on StackOverflow and elsewhere. It is related to the OS, compiler installation and overall the available resources (memory), not hls4ml. You can check the `build_lib.sh` in the output directory to see how it is compiled and you can run it manually while looking for a fix (no need for Jupyter environment).

**Reply by Govarshini99** [reply]
Will check it up , thanks for the reply.

**Reply by Govarshini99** [reply]
I ran build_lib.sh and got this error 
g++: internal compiler error: Segmentation fault signal terminated program cclplus Please submit a full bug report, with preprocessed source if appropriate. See <file:///usr/share/doc/gcc-11/README.Bugs> for instructions. / collect2: error: ld returned 1 exit status usr/bin/ld: cannot find myproject.o: No such file or directory 

Also checked the memory usage when the compilation was run  seems to be full RAM memory is utilized , i tried to increase the swap memory size and reduce the complexity of my model , but still this issue persists...


```
def inverted_residual_block(x, expansion_factor, filters, stride):
    in_channels = x.shape[-1]
    expanded_channels = in_channels * expansion_factor
    # Expansion (1x1 Convolution)
    if expansion_factor != 1:
        x = layers.Conv2D(expanded_channels, kernel_size=1, padding='same', use_bias=False)(x)
        x = layers.BatchNormalization()(x)
        x = layers.ReLU(max_value=6.0)(x)
    
    # 3x3 Convolution (replacing Depthwise Conv2D)
    x = layers.Conv2D(expanded_channels, kernel_size=3, strides=stride, padding='same', use_bias=False)(x)
    x = layers.BatchNormalization()(x)
    x = layers.ReLU(max_value=6.0)(x)
    
    # Pointwise Convolution (1x1)
    x = layers.Conv2D(filters, kernel_size=1, padding='same', use_bias=False)(x)
    x = layers.BatchNormalization()(x)
    
    return x


def build_reduced_mobilenetv2(input_shape, num_classes):
    inputs = layers.Input(shape=input_shape)
    # Initial Convolution Layer 
    x = layers.Conv2D(12, kernel_size=3, strides=2, padding='same', use_bias=False)(inputs)
    x = layers.BatchNormalization()(x)
    x = layers.ReLU(max_value=6.0)(x)
    
    # Fewer and simpler inverted residual blocks
    x = inverted_residual_block(x, expansion_factor=1, filters=12, stride=1)  # Lower expansion
    x = inverted_residual_block(x, expansion_factor=2, filters=16, stride=2)  # Lower filters
    x = inverted_residual_block(x, expansion_factor=2, filters=24, stride=2)  # Final block
    
    # Final Convolution Layer 
    x = layers.Conv2D(320, kernel_size=1, padding='same', use_bias=False)(x)
    x = layers.BatchNormalization()(x)
    x = layers.ReLU(max_value=6.0)(x)
    
    # Global Average Pooling
    x = layers.GlobalAveragePooling2D()(x)
    
    # Fully connected output layer
    outputs = layers.Dense(num_classes, activation='softmax')(x)
    
    # Build the model
    model = models.Model(inputs, outputs)
    
    return model
```

this is the model i'm using

**Reply by Govarshini99** [reply]
Is anybody able to compile this kind of model?? I tried reducing model complexity even more but still facing this issue.

**Reply by ribbles** [reply]
Are you able to get the [quickstart](https://fastmachinelearning.org/hls4ml/intro/setup.html) to work?
